src/backends/perf_pt/collect.c

```diff
@@ -171,8 +171,23 @@ read_aux(void *aux_buf, struct perf_event_mmap_page *hdr,
         new_data_size = (size - tail) + head;
     }
 
-    // For now we simply crash out if we exhaust the buffer.
-    assert(trace->len + new_data_size <= trace->capacity);
+    // Decide if we have space for the new data, and reallocate the trace
+    // storage buffer if more space is required.
+    __u64 required_capacity = trace->len + new_data_size;
+    if (required_capacity > trace->capacity) {
+        // Over-allocate to 2x what we need, checking that the result fits in
+        // the size_t argument of realloc(3).
+        assert(required_capacity < SIZE_MAX / 2);
```

Comment:
What's the point of this? If someone allocates *that* much memory, `realloc` will return `null` (assuming Linux doesn't infinitely overcommit?).

Reply:
The assertion is protecting us against a heap overflow.

`realloc` accepts a `size_t` for the new size, and we are about to multiply a (somewhat) user controlled number by 2 before reallocating that much.

Consider the case where `required_capacity` is (for example) `SIZE_MAX / 2 + 1`.  When we multiply that by 2, it overflows to 2. When we pass that to `realloc` we would end up shrinking the buffer (not growing it) leaving us open to overflowing the buffer later.

Reply:
OK, I get the general idea. But if this is user controlled, it should be an error path, not an `assert`, right?

Reply:
Having thought some more, should we perhaps introduce a new error type to indicate that a trace is too long? That's what this effectively means, and I think it's a scenario we should be able to recover from.

If that's what we want, the question then is how to communicate that meaning back to Rust. At current C code can either indicate an errno or a libipt error number.

Reply:
Ah, yes, we will *definitely* encounter this case: if we're unlucky and e.g. trace the last iteration of a loop (in an interpreter that can't catch this), the trace will just go on and on...

As to how to structure it... dunno ;)

Reply:
Would you be OK with me adding a new error "kind" for special cases like this?

Reply:
I can't quite picture things well enough to know what's best, so I suggest you go ahead, then we'll work things out afterwards.

Reply:
Maybe that's a candidate for a separate PR? Is that OK?

Reply:
Provided it's the next PR, OK. We don't want to allow this one to fester for to long, or it'll get difficult to change.